take() ignored when resuming after Exceptions$ErrorCallbackNotImplemented

I am using the resume() functionality. I have experienced a very strange behavior that I consider one or multiple bugs. It might also be related to https://github.com/rsocket/rsocket-java/issues/857, since it also logs the message mentioned there.

When calling requestStream() and subscribing to the returned flux without an error handler, the connection gets closed with the following error log:

```
02-06-2020 19:20:31.306 [reactor-tcp-nio-2] ERROR r.n.c.ChannelOperationsHandler - [id: 0x8793d78e, L:/127.0.0.1:55222 - R:localhost/127.0.0.1:7890] Error was received while reading the incoming data. The connection will be closed.
reactor.core.Exceptions$ErrorCallbackNotImplemented: ApplicationErrorException (0x201): Some Exception for this test
Caused by: io.rsocket.exceptions.ApplicationErrorException: Some Exception for this test
	at io.rsocket.exceptions.Exceptions.from(Exceptions.java:76)
	at io.rsocket.core.RSocketRequester.handleFrame(RSocketRequester.java:646)
	at io.rsocket.core.RSocketRequester.handleIncomingFrames(RSocketRequester.java:585)
```

When reusing the client connection for another request, the server does receive the request and starts processing it (seen from the logs). However, it does not react on cancel signals from the client, nor does the client receive any results (`next` signals).

See the code below.

## Expected Behavior
After reconnect, `cancel` and `next` signals should still be honored.

## Actual Behavior
They are not. See test code.

## Steps to Reproduce

```java
@Test
public void strangeResumeBehavior() throws Exception {
	CloseableChannel server = RSocketServer.create((setup, sendingRSocket) -> Mono.just(new RSocket() {
		@Override
		public Flux<Payload> requestStream(Payload payload) {
			if (payload.getDataUtf8().equals("flux")) {
				return Flux.interval(Duration.ofMillis(1))
						.doOnNext(aLong -> System.out.println("Server-side doOnNext: " + aLong))
						.doOnCancel(() -> System.out.println("Server-side doOnCancel"))
						.map(aLong -> DefaultPayload.create("" + aLong));
			} else {
				return Flux.error(new RuntimeException("Some Exception for this test"));
			}
		}
	}))
			.resume(new Resume()
					.sessionDuration(Duration.ofSeconds(30))
					.retry(Retry.fixedDelay(Long.MAX_VALUE, Duration.ofSeconds(2))
							.doBeforeRetry(retrySignal -> System.out.println("RETRYING to connect.")))) // never called. why not?
			.bind(TcpServerTransport.create(TcpServer.create().host("localhost").port(7890)))
			.block();

	RSocket client = RSocketConnector.create()
			.resume(new Resume()
					.sessionDuration(Duration.ofSeconds(30))
					.retry(Retry.fixedDelay(Long.MAX_VALUE, Duration.ofSeconds(5))))
			.connect(TcpClientTransport.create(TcpClient.create().host("localhost").port(7890)))
			.retryWhen(Retry.backoff(Long.MAX_VALUE, Duration.ofSeconds(10)))
			.cache()
			.block();


	client.requestStream(DefaultPayload.create("flux"))
			.take(10) // works
			.subscribe(payload -> System.out.println(payload.getDataUtf8()));

	Thread.sleep(1000);

	client.requestStream(DefaultPayload.create("exception"))
			.subscribe(payload -> System.out.println(payload.getDataUtf8()));  // No error handler, so "The connection will be closed."

	Thread.sleep(1000);

	client.requestStream(DefaultPayload.create("flux"))
			.take(10) // <<< Ignored. Why?
			.subscribe(payload -> System.out.println(payload.getDataUtf8())); // <<< Never arrives...

	Thread.sleep(6_000);
}
```

## Possible Solution
<!--- Not obligatory, but you can suggest a fix/reason for the bug. -->

## Your Environment

* RSocket version(s) used: 1.0.0
* Other relevant libraries versions (eg. `netty`, ...): default rsocket dependencies
* Platform (eg. JVM version (`javar -version`) or Node version (`node --version`)):
openjdk 14.0.1 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.1+7)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 14.0.1+7, mixed mode, sharing)
* OS and version (eg `uname -a`): MacOS

--- a/rsocket-core/src/main/java/io/rsocket/core/RSocketRequester.java
+++ b/rsocket-core/src/main/java/io/rsocket/core/RSocketRequester.java
@@ -691,7 +691,22 @@ class RSocketRequester implements RSocket {
           handleMissingResponseProcessor(streamId, type, frame);
           return;
         }
-        receiver.onError(Exceptions.from(streamId, frame));
+
+        // FIXME: when https://github.com/reactor/reactor-core/issues/2176 is resolved
+        //        This is workaround to handle specific Reactor related case when
+        //        onError call may not return normally
+        try {
+          receiver.onError(Exceptions.from(streamId, frame));
+        } catch (RuntimeException e) {
+          if (reactor.core.Exceptions.isBubbling(e)
+              || reactor.core.Exceptions.isErrorCallbackNotImplemented(e)) {
+            if (LOGGER.isDebugEnabled()) {
+              Throwable unwrapped = reactor.core.Exceptions.unwrap(e);
+              LOGGER.debug("Unhandled dropped exception", unwrapped);
+            }
+          }
+        }
+
         receivers.remove(streamId);
         break;
       case CANCEL:

--- a/rsocket-core/src/main/java/io/rsocket/core/RSocketResponder.java
+++ b/rsocket-core/src/main/java/io/rsocket/core/RSocketResponder.java
@@ -338,7 +338,20 @@ class RSocketResponder implements RSocket {
         case ERROR:
           receiver = channelProcessors.get(streamId);
           if (receiver != null) {
-            receiver.onError(io.rsocket.exceptions.Exceptions.from(streamId, frame));
+            // FIXME: when https://github.com/reactor/reactor-core/issues/2176 is resolved
+            //        This is workaround to handle specific Reactor related case when
+            //        onError call may not return normally
+            try {
+              receiver.onError(io.rsocket.exceptions.Exceptions.from(streamId, frame));
+            } catch (RuntimeException e) {
+              if (reactor.core.Exceptions.isBubbling(e)
+                  || reactor.core.Exceptions.isErrorCallbackNotImplemented(e)) {
+                if (LOGGER.isDebugEnabled()) {
+                  Throwable unwrapped = reactor.core.Exceptions.unwrap(e);
+                  LOGGER.debug("Unhandled dropped exception", unwrapped);
+                }
+              }
+            }
           }
           break;
         case NEXT_COMPLETE:
@@ -448,22 +461,10 @@ class RSocketResponder implements RSocket {
             try {
               if (!PayloadValidationUtils.isValid(mtu, payload)) {
                 payload.release();
-                // specifically for requestChannel case so when Payload is invalid we will not be
-                // sending CancelFrame and ErrorFrame
-                // Note: CancelFrame is redundant and due to spec
-                // (https://github.com/rsocket/rsocket/blob/master/Protocol.md#request-channel)
-                // Upon receiving an ERROR[APPLICATION_ERROR|REJECTED|CANCELED|INVALID], the stream
-                // is
-                // terminated on both Requester and Responder.
-                // Upon sending an ERROR[APPLICATION_ERROR|REJECTED|CANCELED|INVALID], the stream is
-                // terminated on both the Requester and Responder.
-                if (requestChannel != null) {
-                  channelProcessors.remove(streamId, requestChannel);
-                }
-                cancel();
                 final IllegalArgumentException t =
                     new IllegalArgumentException(INVALID_PAYLOAD_ERROR_MESSAGE);
-                handleError(streamId, t);
+
+                cancelStream(t);
                 return;
               }
 
@@ -471,20 +472,27 @@ class RSocketResponder implements RSocket {
                   PayloadFrameCodec.encodeNextReleasingPayload(allocator, streamId, payload);
               sendProcessor.onNext(byteBuf);
             } catch (Throwable e) {
-              // specifically for requestChannel case so when Payload is invalid we will not be
-              // sending CancelFrame and ErrorFrame
-              // Note: CancelFrame is redundant and due to spec
-              // (https://github.com/rsocket/rsocket/blob/master/Protocol.md#request-channel)
-              // Upon receiving an ERROR[APPLICATION_ERROR|REJECTED|CANCELED|INVALID], the stream is
-              // terminated on both Requester and Responder.
-              // Upon sending an ERROR[APPLICATION_ERROR|REJECTED|CANCELED|INVALID], the stream is
-              // terminated on both the Requester and Responder.
-              if (requestChannel != null) {
-                channelProcessors.remove(streamId, requestChannel);
-              }
-              cancel();
-              handleError(streamId, e);
+              cancelStream(e);
             }
+          }
+
+          private void cancelStream(Throwable t) {
+            // Cancel the output stream and send an ERROR frame but do not dispose the
+            // requestChannel (i.e. close the connection) since the spec allows to leave
+            // the channel in half-closed state.
+            // specifically for requestChannel case so when Payload is invalid we will not be
+            // sending CancelFrame and ErrorFrame
+            // Note: CancelFrame is redundant and due to spec
+            // (https://github.com/rsocket/rsocket/blob/master/Protocol.md#request-channel)
+            // Upon receiving an ERROR[APPLICATION_ERROR|REJECTED|CANCELED|INVALID], the stream
+            // is terminated on both Requester and Responder.
+            // Upon sending an ERROR[APPLICATION_ERROR|REJECTED|CANCELED|INVALID], the stream is
+            // terminated on both the Requester and Responder.
+            if (requestChannel != null) {
+              channelProcessors.remove(streamId, requestChannel);
+            }
+            cancel();
+            handleError(streamId, t);
           }
 
           @Override
@@ -502,8 +510,7 @@ class RSocketResponder implements RSocket {
               // Note: CancelFrame is redundant and due to spec
               // (https://github.com/rsocket/rsocket/blob/master/Protocol.md#request-channel)
               // Upon receiving an ERROR[APPLICATION_ERROR|REJECTED|CANCELED|INVALID], the stream
-              // is
-              // terminated on both Requester and Responder.
+              // is terminated on both Requester and Responder.
               // Upon sending an ERROR[APPLICATION_ERROR|REJECTED|CANCELED|INVALID], the stream is
               // terminated on both the Requester and Responder.
               if (requestChannel != null && !requestChannel.isDisposed()) {

--- a/rsocket-core/src/test/java/io/rsocket/core/RSocketRequesterTest.java
+++ b/rsocket-core/src/test/java/io/rsocket/core/RSocketRequesterTest.java
@@ -1081,6 +1081,28 @@ public class RSocketRequesterTest {
     }
   }
 
+  @Test
+  // see https://github.com/rsocket/rsocket-java/issues/858
+  public void testWorkaround858() {
+    ByteBuf buffer = rule.alloc().buffer();
+    buffer.writeCharSequence("test", CharsetUtil.UTF_8);
+
+    rule.socket.requestResponse(ByteBufPayload.create(buffer)).subscribe();
+
+    rule.connection.addToReceivedBuffer(
+        ErrorFrameCodec.encode(rule.alloc(), 1, new RuntimeException("test")));
+
+    Assertions.assertThat(rule.connection.getSent())
+        .hasSize(1)
+        .first()
+        .matches(bb -> FrameHeaderCodec.frameType(bb) == REQUEST_RESPONSE)
+        .matches(ByteBuf::release);
+
+    Assertions.assertThat(rule.socket.isDisposed()).isFalse();
+
+    rule.assertHasNoLeaks();
+  }
+
   public static class ClientSocketRule extends AbstractSocketRule<RSocketRequester> {
     @Override
     protected RSocketRequester newRSocket() {

--- a/rsocket-core/src/test/java/io/rsocket/core/RSocketResponderTest.java
+++ b/rsocket-core/src/test/java/io/rsocket/core/RSocketResponderTest.java
@@ -429,23 +429,23 @@ public class RSocketResponderTest {
 
       ByteBuf requestNFrame = RequestNFrameCodec.encode(allocator, 1, Integer.MAX_VALUE);
 
-        ByteBuf m1 = allocator.buffer();
-        m1.writeCharSequence("m1", CharsetUtil.UTF_8);
-        ByteBuf d1 = allocator.buffer();
-        d1.writeCharSequence("d1", CharsetUtil.UTF_8);
-        Payload np1 = ByteBufPayload.create(d1, m1);
+      ByteBuf m1 = allocator.buffer();
+      m1.writeCharSequence("m1", CharsetUtil.UTF_8);
+      ByteBuf d1 = allocator.buffer();
+      d1.writeCharSequence("d1", CharsetUtil.UTF_8);
+      Payload np1 = ByteBufPayload.create(d1, m1);
 
-        ByteBuf m2 = allocator.buffer();
-        m2.writeCharSequence("m2", CharsetUtil.UTF_8);
-        ByteBuf d2 = allocator.buffer();
-        d2.writeCharSequence("d2", CharsetUtil.UTF_8);
-        Payload np2 = ByteBufPayload.create(d2, m2);
+      ByteBuf m2 = allocator.buffer();
+      m2.writeCharSequence("m2", CharsetUtil.UTF_8);
+      ByteBuf d2 = allocator.buffer();
+      d2.writeCharSequence("d2", CharsetUtil.UTF_8);
+      Payload np2 = ByteBufPayload.create(d2, m2);
 
-        ByteBuf m3 = allocator.buffer();
-        m3.writeCharSequence("m3", CharsetUtil.UTF_8);
-        ByteBuf d3 = allocator.buffer();
-        d3.writeCharSequence("d3", CharsetUtil.UTF_8);
-        Payload np3 = ByteBufPayload.create(d3, m3);
+      ByteBuf m3 = allocator.buffer();
+      m3.writeCharSequence("m3", CharsetUtil.UTF_8);
+      ByteBuf d3 = allocator.buffer();
+      d3.writeCharSequence("d3", CharsetUtil.UTF_8);
+      Payload np3 = ByteBufPayload.create(d3, m3);
 
       FluxSink<Payload> sink = sinks[0];
       RaceTestUtils.race(
@@ -464,9 +464,10 @@ public class RSocketResponderTest {
 
       Assertions.assertThat(rule.connection.getSent()).allMatch(ReferenceCounted::release);
 
-      assertSubscriber.assertTerminated()
-                      .assertError(CancellationException.class)
-                      .assertErrorMessage("Disposed");
+      assertSubscriber
+          .assertTerminated()
+          .assertError(CancellationException.class)
+          .assertErrorMessage("Disposed");
       Assertions.assertThat(assertSubscriber.values()).allMatch(ReferenceCounted::release);
       rule.assertHasNoLeaks();
     }
@@ -770,6 +771,42 @@ public class RSocketResponderTest {
 
   private static Stream<FrameType> refCntCases() {
     return Stream.of(REQUEST_RESPONSE, REQUEST_STREAM, REQUEST_CHANNEL);
+  }
+
+  @Test
+  // see https://github.com/rsocket/rsocket-java/issues/858
+  public void testWorkaround858() {
+    ByteBuf buffer = rule.alloc().buffer();
+    buffer.writeCharSequence("test", CharsetUtil.UTF_8);
+
+    TestPublisher<Payload> testPublisher = TestPublisher.create();
+
+    rule.setAcceptingSocket(
+        new RSocket() {
+          @Override
+          public Flux<Payload> requestChannel(Publisher<Payload> payloads) {
+            Flux.from(payloads).doOnNext(ReferenceCounted::release).subscribe();
+
+            return testPublisher.flux();
+          }
+        });
+
+    rule.connection.addToReceivedBuffer(
+        RequestChannelFrameCodec.encodeReleasingPayload(
+            rule.alloc(), 1, false, 1, ByteBufPayload.create(buffer)));
+    rule.connection.addToReceivedBuffer(
+        ErrorFrameCodec.encode(rule.alloc(), 1, new RuntimeException("test")));
+
+    Assertions.assertThat(rule.connection.getSent())
+        .hasSize(1)
+        .first()
+        .matches(bb -> FrameHeaderCodec.frameType(bb) == REQUEST_N)
+        .matches(ReferenceCounted::release);
+
+    Assertions.assertThat(rule.socket.isDisposed()).isFalse();
+    testPublisher.assertWasCancelled();
+
+    rule.assertHasNoLeaks();
   }
 
   public static class ServerSocketRule extends AbstractSocketRule<RSocketResponder> {

